### PR TITLE
[FA3 Feature] Support multi modal Llama-3.2-11B-Vision-Instruct

### PR DESCRIPTION
**Pull request by zcnrex**
## Motivation
#4709 

Support `Llama-3.2-11B-Vision-Instruct` in the FlashAttention 3 backend

```
python3 -m sglang.launch_server --model-path meta-llama/Llama-3.2-11B-Vision-Instruct --chat-template=llama_3_vision --attention-backend fa3
```

As detailed below, overall we see consistent speed and accuracy improvement in FA3 vs Flashinfer.

## MMMU Benchmark
official:
https://huggingface.co/meta-llama/Llama-3.2-11B-Vision

#### fa3
```
python benchmark/mmmu/bench_sglang.py
Benchmark time: 374.2589066028595
Overall accuracy: 0.346
```

#### flashinfer
```
python benchmark/mmmu/bench_sglang.py
Benchmark time: 383.1065626144409
Overall accuracy: 0.325
```

#### fa3
```
python benchmark/mmmu/bench_sglang.py --parallel 1
Benchmark time: 545.3505992889404
Overall accuracy: 0.367

python benchmark/mmmu/bench_sglang.py
Benchmark time: 547.4442737102509
Overall accuracy: 0.348
```

#### flashinfer
```
python benchmark/mmmu/bench_sglang.py --parallel 1
Benchmark time: 565.195394039154
Overall accuracy: 0.342

python benchmark/mmmu/bench_sglang.py
Benchmark time: 578.095338344574
Overall accuracy: 0.348
```

#### fa3
```
python3 /home/jobuser/sglang/python/sglang/test/send_one.py --image
 A man in a yellow jacket is ironing clothes on the back of a yellow taxi cab. The taxi is driving down a city street. The man is standing on the sidewalk, holding the ironing board and iron, and appears to be ironing a shirt. The taxi has a white sign on the back that says "Taxi" in black letters. The background is a city street with tall buildings and trees. The overall atmosphere suggests a busy urban setting, with the man's unusual activity adding a touch of humor and surprise to the scene. The image captures a moment of everyday life in the city, highlighting the resourcefulness and adaptability of the man in the yellow jacket. The image conveys a sense of urban bustle and activity, with the taxi and the man's ironing board adding to the sense of movement and energy. The image also suggests a sense of community and connection, as the man is ironing in a public space, surrounded by other people and vehicles. The image is a commentary on the human ability to find ways to get things done, even in unexpected places and situations. It highlights the resourcefulness and adaptability of the man in the yellow jacket, who is able to find a way to iron his clothes in the middle of the street. The image is a celebration of the human spirit and the ability to find joy and humor in everyday situations. The image is a commentary on the urban environment and the ways in which people interact with their surroundings. It highlights the beauty and complexity of city life, where people from all walks of life come together to create a vibrant and dynamic community. The image is a celebration of the human experience and the ways in which we find meaning and purpose in our daily lives. The image is a commentary on the role of technology in our lives, and how it can be used to make our lives easier and more convenient. The image is a celebration of the human ability to find beauty and joy in unexpected places, and to make the most of every situation. The image is a commentary on the importance of community and connection in our lives, and the ways in which we can come together to support and uplift one another. The image is a celebration of the human spirit and the ability to find happiness and fulfillment in the midst of chaos and uncertainty. The image is a commentary on the role of creativity and resourcefulness in our lives, and the ways in which we can use our imagination and ingenuity to overcome obstacles and achieve our goals. The image is a celebration of the human ability to find beauty and

acc_length=1.00
speed=101.20 token/s
```

#### flashinfer
```
 A man in a yellow jacket is ironing clothes on the back of a yellow taxi cab. The taxi is driving down a city street. The man is standing on the sidewalk, holding the ironing board and iron, and appears to be ironing a shirt. The taxi has a white license plate with black lettering that reads "TAXI" and a yellow sign on the side that says "TAXI." The background is a city street with tall buildings and trees. The overall atmosphere suggests a busy urban setting, with the man's unusual activity adding a touch of humor and surprise to the scene. The image captures a moment of everyday life in the city, highlighting the resourcefulness and adaptability of the man in the yellow jacket. The image conveys a sense of urban bustle and activity, with the taxi and the man's ironing board adding to the sense of movement and energy. The image also suggests a sense of community and connection, as the man is ironing in a public space, surrounded by other people and vehicles. The image is a commentary on the human experience, highlighting the ways in which people adapt and find ways to make do in unexpected situations. The image is a celebration of the human spirit, showing how people can find joy and purpose in even the most mundane tasks. The image is a commentary on the urban environment, highlighting the ways in which cities can be both chaotic and beautiful. The image is a reflection of the human condition, showing how people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a celebration of the human experience, highlighting the ways in which people can find beauty and joy in the everyday. The image is a commentary on the human condition, showing how people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a reflection of the human experience, showing how people can find beauty and joy in the everyday. The image is a celebration of the human experience, highlighting the ways in which people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a commentary on the human condition, showing how people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a reflection of the human experience, showing how people can find beauty and joy in the everyday. The image is a celebration of the human experience, highlighting the ways in which people can find meaning and purpose in their daily lives,

acc_length=1.00
speed=99.21 token/s
```

#### fa3
```
 A man in a yellow jacket is ironing clothes on the back of a yellow taxi cab. The taxi is driving down a city street. The man is standing on the sidewalk, holding the ironing board and iron, and appears to be ironing a shirt. The taxi has a white sign on the back that says "Taxi" in black letters. The background is a city street with tall buildings and trees. The overall atmosphere suggests a busy urban setting, with the man's unusual activity adding a touch of humor and surprise to the scene. The image captures a moment of everyday life in the city, highlighting the resourcefulness and adaptability of the man in the yellow jacket. The image conveys a sense of urban bustle and activity, with the taxi and the man's ironing board adding to the sense of movement and energy. The image also suggests a sense of community and connection, as the man is ironing in a public space, surrounded by other people and vehicles. The image is a commentary on the human experience, highlighting the ways in which people find ways to make do with what they have, even in the midst of chaos and activity. The image is a celebration of the human spirit, showing how people can find joy and purpose in the most unexpected places. The image is a commentary on the urban environment, highlighting the ways in which cities can be both chaotic and beautiful, and the ways in which people can find beauty in the everyday. The image is a reflection of the human condition, showing how people can find meaning and purpose in the most mundane tasks, and how they can turn even the most ordinary moments into something extraordinary. The image is a celebration of the human experience, highlighting the ways in which people can find joy and beauty in the most unexpected places, and the ways in which they can turn even the most ordinary moments into something extraordinary. The image is a commentary on the human condition, showing how people can find meaning and purpose in the most mundane tasks, and how they can turn even the most ordinary moments into something extraordinary. The image is a reflection of the human experience, highlighting the ways in which people can find joy and beauty in the most unexpected places, and the ways in which they can turn even the most ordinary moments into something extraordinary. The image is a celebration of the human experience, highlighting the ways in which people can find joy and beauty in the most unexpected places, and the ways in which they can turn even the most ordinary moments into something extraordinary. The image is a commentary on the human condition, showing

acc_length=1.00
speed=58.75 token/s
```

#### flashinfer
```
 A man in a yellow jacket is ironing clothes on the back of a yellow taxi cab. The taxi is driving down a city street. The man is standing on the sidewalk, holding the ironing board and iron, and appears to be ironing a shirt. The taxi has a white license plate with black lettering that reads "TAXI" and a yellow sign on the side that says "TAXI." The background is a city street with tall buildings and trees. The overall atmosphere suggests a busy urban setting, with the man's unusual activity adding a touch of humor and surprise to the scene. The image captures a moment of everyday life in the city, highlighting the resourcefulness and adaptability of the man in the yellow jacket. The image conveys a sense of urban bustle and activity, with the taxi and the man's ironing board adding to the sense of movement and energy. The image also suggests a sense of community and connection, as the man is ironing in a public space, surrounded by other people and vehicles. The image is a commentary on the human experience, highlighting the ways in which people adapt and find ways to make do in unexpected situations. The image is a celebration of the human spirit, showing how people can find joy and purpose in even the most mundane tasks. The image is a commentary on the urban environment, highlighting the ways in which cities can be both chaotic and beautiful. The image is a reflection of the human condition, showing how people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a celebration of the human experience, highlighting the ways in which people can find beauty and joy in the everyday. The image is a commentary on the human condition, showing how people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a reflection of the human experience, showing how people can find beauty and joy in the everyday. The image is a celebration of the human experience, highlighting the ways in which people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a commentary on the human condition, showing how people can find meaning and purpose in their daily lives, even in the midst of chaos and uncertainty. The image is a reflection of the human experience, showing how people can find beauty and joy in the everyday. The image is a celebration of the human experience, highlighting the ways in which people can find meaning and purpose in their daily lives,

acc_length=1.00
speed=43.15 token/s
```

## Checklist

TODO

- [ ] Add encoder batch size > 1 support if needed
- [ ] Support page size > 1

### CODE REVIEW

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 1]
```diff
@@ -203,6 +213,31 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+                (1, 0),
+            )
+            metadata.encoder_max_seq_len_k = metadata.encoder_lens_int32.max().item()
+            metadata.encoder_cu_seqlens_q = metadata.cu_seqlens_q
+            metadata.encoder_max_seq_len_q = metadata.max_seq_len_q
+            metadata.encoder_page_table = forward_batch.req_to_token_pool.req_to_token[
+                forward_batch.req_pool_indices, : metadata.encoder_max_seq_len_k
+            ]
+            ## TODO: support batch size > 1
+            metadata.page_table = forward_batch.req_to_token_pool.req_to_token[
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/flashattention_backend.py:465]: metadata.encoder_page_table?

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 2]
```diff
@@ -250,11 +285,11 @@ def forward_extend(
         # here is two side inclusive
         window_size = (
             (layer.sliding_window_size, 0)
-            if layer.sliding_window_size is not None
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/flashattention_backend.py:486]: assert not (
            model_runner.sliding_window_size is not None
            and model_runner.model_config.is_encoder_decoder
        ), "Sliding window and cross attention are not supported together"
(copied from FlashInfer Backend)

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 1 above)
Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py:465]: This is because the text in the page table is after the image.

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 2 above)
Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py:486]: Thanks will test more. I saw (-1, 0) during dev, was really want to make sure it's (-1, -1)

Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py:486]: Removed, thanks.

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 3]
```diff
@@ -203,6 +213,40 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+            for i, seq_len in enumerate(metadata.cache_seqlens_int32):
+                metadata.page_table[:, :seq_len] = (
+                    forward_batch.req_to_token_pool.req_to_token[
+                        forward_batch.req_pool_indices,
+                        metadata.encoder_lens_int32[i] : (
+                            metadata.encoder_lens_int32[i] + seq_len
+                        ),
+                    ]
+                )
+                metadata.page_table[:, seq_len:].fill_(0)
```
Comment (qingquansong) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: I think we probably don't need to fill zero here now since the `encoder_cu_seqlens_k` and `cache_seqlens_int32` can tell us the length of each request. Thank you!

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 4]
```diff
@@ -203,6 +213,33 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+            metadata.encoder_cu_seqlens_k = torch.nn.functional.pad(
+                torch.cumsum(metadata.encoder_lens_int32, dim=0, dtype=torch.int32),
+                (1, 0),
+            )
+            metadata.encoder_max_seq_len_k = metadata.encoder_lens_int32.max().item()
+            metadata.encoder_page_table = forward_batch.req_to_token_pool.req_to_token[
+                forward_batch.req_pool_indices, : metadata.encoder_max_seq_len_k
+            ]
+
+            # TODO: support len(forward_batch.encoder_lens) > 1
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: i think you already supported it?

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 5]
```diff
@@ -203,6 +213,33 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
                 metadata.cu_seqlens_q = metadata.cu_seqlens_k
                 metadata.max_seq_len_q = metadata.max_seq_len_k
 
+        if forward_batch.encoder_lens is not None:
+            assert (
+                len(forward_batch.encoder_lens) == 1
+            ), "Only encoder batch size 1 is supported for now"
+
+            metadata.encoder_lens_int32 = (
+                forward_batch.encoder_lens.clone()
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: curious why we need to clone and detach?

Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: `encoder_lens` isn't Tensor, but a list of ints. It threw error when directly copy. Will investigate how to improve if possible.

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 4 above)
Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: Would need to construct a correct page table for text inputs when `len(forward_batch.encoder_lens) > 1`.

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 3 above)
Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: Thanks removed

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 5 above)
Comment (zcnrex) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: Actually confused it with `encoder_len_cpu`. Changed to directly assign.
